users/app.py

```diff
@@ -37,20 +41,60 @@ def get_authorization():
     return jsonify(edit_access=authorized)
 
 
-@app.route('/v1/', methods=['PUT'])
-def add_update_user():
-    """Add or update the user in the db and returns new user object."""
-    user = request.getJSON()
-    if user is None:
-        # TODO(mukobi) validate the user object has everything it needs
-        return jsonify(error="You must supply a valid user in the body")
-    # TODO(mukobi) add or update the user in the database
-    added_new_user = True
-    # TODO(mukobi) get the new user object from the db to return
-    user_object = {
-        "user_id": "0", "username": "Dummy User", "edit_access": False
-    }
-    return jsonify(user_object), (201 if added_new_user else 200)
+@app.route('/v1/authenticate', methods=['POST'])
+def authenticate_user():
+    """Authenticate user, upsert in the db, and return new user object.
+
+    Request data:
+        'gauth_token': Google ID token to authenticate.
+
+    Response:
+        201: user object as it is in the db if authentication
+            was successful.
+        400: error message if authentication was not successful.
+    """
+    gauth_token = request.form.get('gauth_token')
+    if gauth_token is None:
+        return "Error: You must authenticate through Google.", 400
+    try:
+        idinfo = get_user_from_gauth_token(gauth_token)
+        user_object = {
+            "user_id": idinfo["sub"],
+            "name": idinfo["name"],
```

Comment:
Maybe add a test where `"name"` isn't given - in the case where the user doesn't give the correct permissions it might not be returned.

Reply:
I'm not too sure how possible that is, since I think the smallest permission scope is 'profile' which contains an ID and a name, whereas things like email are additional scopes that aren't guaranteed. In any case, I added a test at https://github.com/knative-portability/large-events/pull/31/commits/62a4902f21f8190f09663d7d31b658d85cb2b807